src/boot/efi/linux.c

```diff
@@ -4,89 +4,197 @@
+   see https://github.com/torvalds/linux/blob/v5.13/drivers/firmware/efi/libstub/efi-stub-helper.c
+ */
+static const struct {
+        VENDOR_DEVICE_PATH vendor;
+        EFI_DEVICE_PATH end;
+} _packed_ efi_initrd_device_path = {
+        .vendor = {
+                {
+                        MEDIA_DEVICE_PATH,
+                        MEDIA_VENDOR_DP,
+                        sizeof(efi_initrd_device_path.vendor)
+                },
+                LINUX_INITRD_MEDIA_GUID
+        },
+        .end = {
+                END_DEVICE_PATH_TYPE,
+                END_ENTIRE_DEVICE_PATH_SUBTYPE,
+                sizeof(efi_initrd_device_path.end)
+        }
+};
+
+#pragma GCC diagnostic pop
+
+EFIAPI EFI_STATUS initrd_load_file(
+                EFI_LOAD_FILE_PROTOCOL *this,
+                EFI_DEVICE_PATH *file_path,
+                BOOLEAN boot_policy,
+                UINTN *buffer_size,
+                VOID *buffer
+);
+
+static const EFI_LOAD_FILE_INTERFACE load_file2_interface = {
+        .LoadFile = initrd_load_file
+};
+
+EFIAPI EFI_STATUS initrd_load_file(
+                EFI_LOAD_FILE_PROTOCOL* this,
+                EFI_DEVICE_PATH* file_path,
+                BOOLEAN boot_policy,
+                UINTN* buffer_size,
+                VOID* buffer) {
+
+        if (this != &load_file2_interface || !buffer_size || !file_path)
+                return EFI_INVALID_PARAMETER;
+        if (boot_policy)
+                return EFI_UNSUPPORTED;
+
+        if (!_initrd_length || !_initrd_buffer)
+                return EFI_NOT_FOUND;
+
+        if (!buffer || *buffer_size < _initrd_length) {
+                *buffer_size = _initrd_length;
+                return EFI_BUFFER_TOO_SMALL;
+        }
+
+        CopyMem(buffer, _initrd_buffer, _initrd_length);
+        *buffer_size = _initrd_length;
+        return EFI_SUCCESS;
+}
+
+static EFI_STATUS initrd_register(
+                VOID *initrd_buffer,
+                UINTN initrd_length) {
+
+        EFI_STATUS err;
+        EFI_DEVICE_PATH *dp = (EFI_DEVICE_PATH *) &efi_initrd_device_path;
+        EFI_HANDLE handle;
+
+        if (!initrd_buffer || !initrd_length)
+                return EFI_SUCCESS;
+
+        err = uefi_call_wrapper(BS->LocateDevicePath, 3, &EfiLoadFile2Protocol , &dp, &handle);
+
+        // InitrdMedia is already registered
+        if (err != EFI_NOT_FOUND)
+                return EFI_ALREADY_STARTED;
+
+
+        _initrd_buffer = initrd_buffer;
+        _initrd_length = initrd_length;
+
+        // crate a new handle and register the LoadFile2 protocol with the InitrdMediaPath on it
+        err = uefi_call_wrapper(
+                                BS->InstallMultipleProtocolInterfaces, 6,
+                                &_initrd_handle,
+                                &DevicePathProtocol, &efi_initrd_device_path,
+                                &EfiLoadFile2Protocol, &load_file2_interface,
+                                NULL);
+
+        return err;
```

Comment:
you can directly do `return uefi_call_wrapper(…` — no need to go indirect via `err`.

If this call fails, then we'll leave things half initialized, i.e. the static vars will already be set. I think it would make sense to avoid that, and sure we leave either everyting initialized or nothing, but not half-initialized state. Simply as a matter of hygiene, even if it doesn#t matter right now.